src/singularity.ts

```diff
@@ -1622,34 +1657,33 @@ export const singularityPerks: SingularityPerk[] = [
       } else {
         return i18next.t('singularity.perks.immaculateAlchemy.default')
       }
-    }
+    },
+    ID: 'immaculateAlchemy'
   },
   {
     name: () => {
       return i18next.t('singularity.perkNames.skrauQ')
     },
     levels: [200],
     description: () => {
-      const amt = Math.pow((player.singularityCount - 179) / 20, 2)
+      const amt = format(Math.pow((player.singularityCount - 179) / 20, 2), 4)
       return i18next.t('singularity.perks.skrauQ', { amt })
-    }
+    },
+    ID: 'skrauQ'
   }
 ]
 
 export const updateSingularityPerks = (): void => {
   const singularityCount = player.highestSingularityCount
-
   const strH = i18next.t('singularity.perks.header', {
     ord: toOrdinal(singularityCount)
   })
-  // TODO: Take current perk description and display it here.
-  const strD = i18next.t('singularity.perks.description')
-
-  const str = `${getAvailablePerksDescription(singularityCount)}`
 
   DOMCacheGetOrSet('singularityPerksHeader').innerHTML = strH
-  DOMCacheGetOrSet('singularityPerksDesc').innerHTML = strD
-  DOMCacheGetOrSet('singularityPerksGrid').innerHTML = str
+  DOMCacheGetOrSet('singularityPerksText').innerHTML = `${i18next.t('general.level')} # - (Singularity #)`
+  DOMCacheGetOrSet('singularityPerksDesc').innerHTML = i18next.t('singularity.perks.description')
+  DOMCacheGetOrSet('singularityPerksGrid').innerHTML = `${getAvailablePerksDescription(singularityCount)}`
+  addPerkEventHandlers()
```

Comment:
if `updateSingularityPerks` is called more than once this will create a memory leak.

Reply:
Part of the issue is that the perk menu seems to recreate itself whenever the Singularity tab is visited, where abouts would be a better place to put this function given this?

Reply:
As long as this function is called only after any sets to 'singularityPerksGrid' innerHTML, shoudn't be any memory leaks (that if we assume everyone uses modern browsers that garbage collects no longer connected events), that's also another reason for not using DOMCacheGetOrSet, we want to address new element with same ID, while DOM only allows to address old one (also would cause memory leak since events would pile up on top of each other)